meetings/http/2017-03-14/minutes.md

```diff
@@ -1,24 +1,65 @@
 # HTTP stream meeting 2
 
-### Time: Wednesday March 15th 
-* 6pm CET 
-* 5pm UK 
-* 1pm EST 
-* 10am PST
-            
-            
-### WebEx and Agenda/Minutes doc:
-            
-WebEx:      URL:        https://ibm2.webex.com/ibm2-en/j.php?MTID=m59d2772a43f192bf64ae3bdbd5c77faf
-            password:   swift-http
-        
-Agenda:     https://docs.google.com/document/d/1GTm7mJH2vCAK-Ts4M3t8C3boJ18XqiFnCRuyp55eKJE/edit?usp=sharing
-
-If you'd like to add anything to the agenda ahead of time, please update the agenda document in Google Docs.
 
 ### Initial Agenda:
-* Review of draft HTTP API Proposal 
+* Review of draft HTTP API Pitch 
+  * https://docs.google.com/document/d/1ErHA8RXLiTGvOPHNY0KtbWCvwR1IUz_2ffi7CxCUo_4/edit
 * Discussion of key points:
   * Use of URLRequest/Response
   * Value Types vs. Reference Types
 
+### Attendees:
+* Chris Bailey
+* Ben Cohen
+* Carl Brown
+* Stevey Brown
+* Tony Parker
+* Gregor Milos
+* Shmuel Kallner
+* Alex Blewitt
+* Allen
+* Chuck Goddard
+* George
+* Johannes Weiss
+* Ludovic Dewailly
+* Marcin Kliks
+* Nasser Ebrahim
+* Steve Algernon
+* Michael Papp
+* Thomas Catterall
+
+### Minutes:
+
+_Use of value types vs. reference types for request and response_ 
+* General consensus is in favour of using a value types for request and response.
+  * The argument that the should be references as they are passed around and mutated really don’t hold
+  * It should be noted that the reason why there is still NSURLRequest/Response (reference types) was because a subclass exists, rather than the intend that they should be reference types.
+  * The use of value types will help with reference cycles in closures (like middlewares) which you otherwise need to use weak self for.
+  * Should also help with reducing some of the ARC overhead
```

Comment:
This is all nice and right, but you are mixing up ARC with `UnsafePointer<struct>`. I'm 100% in for using a zero copy architecture, but in Swift that would require the latter (no variable sized structs or objects). ARC has a non-negible synchronisation overhead, it is not the kind of issue the compiler can just fix for you. In contrast (depending on the actual ABI) the compiler *can* reduce copying of structs (if it's all 'let's, there is no reason to copy anything, while the compiler can't know which threads have a retain on an object).

Please be careful with statements like that. There is way too much 'less likely' and 'shouldn't be' for my taste ;-)
To provide an actual example you can profile, check out this [http-c-vs-swift](https://github.com/helje5/http-c-vs-swift). ARC is a huge overhead in that, and the example doesn't even do a lot of referencing back-n-forth :-)

Using objects also has the disadvantage that they need be heap allocated (not sure whether that is strictly true in Swift :-), but I assume it will stay that way). Often times that can be yet another major performance issue which can be avoided with stack allocated structs. (e.g. very little of the data of the HTTP message head should/can be zero copied)

Finally your example refers to net map which is mainly concerned about the block I/O which relates to the design of the stream part of the request, *not* the HTTP message heads. Yes, we absolutely should not copy body parts of the HTTP stream (and I'm still a little more interested how the stream API will look like).

In summary I guess my opinion is that: 
- What the HTTP Swift library passes out is likely going to be wrapped, either a custom "IncomingRequest" object, or a sideline object similar to `WOContext`.
- I don't care too much whether the request is done as a struct or object, my gut feeling is that the struct will be more performant (for the low-level data holder, *not* for the object passed around at high levels). Just guesswork, needs to be tested.
- I think it his highly desirable that client and server share the same infrastructure
- It may be desirable to have a protocol for Request/Response which all Swift server frameworks can share.
- Unfortunately I'm pretty sure this is going to be just the 'default implementation', performance oriented frameworks will do their own thing anyways. Parsing HTTP headers has a lot of optimisation potential, particularly in Swift.